### PR TITLE
Better error handling

### PR DESCRIPTION
**Pull request by tylerdavis**
Need another gut check for style and whatever before I submit the real pull request.  Any and all comments welcome and encouraged.  Thank you.

@mrhamric @detkin @scootklein @duane

### CODE REVIEW

File: lib/hipchat/errors.rb [hunk 1]
```diff
@@ -1,14 +1,46 @@
...
+  class TooManyRequests         < ServiceError;   end
+
+  class ErrorHandler
+
+    # Pass-through to catch error states and raise their appropriate exceptions
+    # @param klass {Symbol} The class of object we are handling the error for
+    # @param identifier {String} An identifying string for the object causing the error
+    # @param response {HTTParty::Response} The HTTParty response/request object
+    def self.catch_and_raise_exception_for(klass, identifier, response)
+      not_found_exception = Module.const_get(HipChat.to_s.to_sym).const_get("Unknown#{klass.capitalize}".to_sym)
```
Comment (detkin) [inline lib/hipchat/errors.rb:29]: This is a little hard to decode exactly what's happening here and why. At least add some doco for this.

File: lib/hipchat/errors.rb [hunk 2]
```diff
@@ -1,14 +1,46 @@
...
+  class UnknownWebhook          < ObjectNotFound; end
+  class TooManyRequests         < ServiceError;   end
+
+  class ErrorHandler
+
+    # Pass-through to catch error states and raise their appropriate exceptions
+    # @param klass {Symbol} The class of object we are handling the error for
+    # @param identifier {String} An identifying string for the object causing the error
+    # @param response {HTTParty::Response} The HTTParty response/request object
+    def self.catch_and_raise_exception_for(klass, identifier, response)
```
Comment (detkin) [inline lib/hipchat/errors.rb]: Maybe the method should be called?

```
response_code_to_exception_for
```

File: spec/hipchat_api_v1_spec.rb [hunk 3]
```diff
@@ -48,7 +48,7 @@
         OpenStruct.new(:code => 403)
       }
 
-      expect { room.history }.to raise_error(HipChat::UnknownResponseCode)
+      expect { room.history }.to raise_error(HipChat::Unauthorized)
```
Comment (detkin) [inline spec/hipchat_api_v1_spec.rb:51]: Doesn't this change the contract of the Gem? Does Unauthorized extend from UnknownResponseCode?

File: lib/hipchat/client.rb [hunk 4]
```diff
@@ -118,13 +102,10 @@ def _rooms
...
-        response[@api.rooms_config[:data_key]].map do |r|
-          HipChat::Room.new(@token, r.merge(:api_version => @api_version, :server_url => @options[:server_url]))
-        end
-      else
-        raise UnknownResponseCode, "Unexpected #{response.code} for room"
+
+      ErrorHandler.response_code_to_exception_for :room, nil, response
+      response[@api.rooms_config[:data_key]].map do |r|
+        HipChat::Room.new(@token, r.merge(:api_version => @api_version, :server_url => @options[:server_url]))
       end
```
Comment (mrhamric) [inline lib/hipchat/client.rb:109]: `api_version: @api_version`, `server_url: @opt...`

File: lib/hipchat/client.rb [hunk 5]
```diff
@@ -136,13 +117,10 @@ def _users
...
-      when 200
-        response[@api.users_config[:data_key]].map do |u|
-          HipChat::User.new(@token, u.merge(:api_version => @api_version, :server_url => @options[:server_url]))
-        end
-      else
-        raise UnknownResponseCode, "Unexpected #{response.code} for user"
+
+      ErrorHandler.response_code_to_exception_for :user, nil, response
+      response[@api.users_config[:data_key]].map do |u|
+        HipChat::User.new(@token, u.merge(:api_version => @api_version, :server_url => @options[:server_url]))
```
Comment (mrhamric) [inline lib/hipchat/client.rb:123]: see previous comment, hashrocket purge!

**Comment by mrhamric** [comment]
lot of red, nice :+1:

**Pull request closed without merging**